---
title: "Tumor tissue and organoids - MicroArray QC and preprocessing"
author: "Hilmar Berger"
output: 
 html_document:
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: true
    code_folding: hide
    
pdf_document:
    fig_caption: true
date: "`r format(Sys.time(), '%d %B, %Y %H:%M:%S')`"
---

```{r, results='hide', message=FALSE}
rm(list=ls())
library(limma)
library(xlsx)
library(pheatmap)
library(xtable)
library(knitr)
```

# Data QC and preprocessing for microarray data

```{r,   message=FALSE, warning=FALSE}
ed = read.xlsx("ExpDesign.xls", 1, stringsAsFactors=F)
ed$sample_ID=as.character(ed$Sample.Number)
ed$ShortID = ifelse(ed$DS_TumorOrganoids, paste(ed$Condition, ed$Site, ed$patient), paste(ed$Condition, ed$Site, ed$patient, ed$Medium))
rownames(ed) = gsub(".txt","",ed$filename)

setwd("../../Data/Raw/")
agilent.datacolumns=list(E='gMedianSignal',Eb = 'gBGMedianSignal',isNonUniform='gIsFeatNonUnifOL', isPopOutlier='gIsFeatPopnOL');
intensities =read.maimages(paste(ed$filename,sep=""), source="agilent.median", green.only=TRUE, columns=agilent.datacolumns)
setwd("../../Code/GeneExpressionMicroarray/")

# fix outdated chip annotations
new_anno_file = "../../Data/Raw/Agilent_14850_annotations_2017-02-08.Rdata"
load(new_anno_file)
old_anno = intensities$genes
take_over_cols = colnames(old_anno)[!colnames(old_anno) %in% c("GeneName","Description","SystematicName")]
tmp = old_anno[,take_over_cols]
tmp$index=1:nrow(tmp)
tmp = merge(tmp, anno_tab_14850, by.x="ProbeName", by.y="ProbeID", all.x=T, sort=F)
new_col_order = c(take_over_cols, colnames(tmp)[!colnames(tmp) %in% take_over_cols])
new_col_order = new_col_order[!new_col_order %in% c("GO_BP","GO_CC","GO_MF")]
new_anno = tmp[order(tmp$index),new_col_order]

intensities$genes = new_anno
```


```{r,  message=FALSE}
intensities$probe_exclude=(intensities$isNonUniform>0)|(intensities$isPopOutlier>0)
intensities$E[intensities$probe_exclude] <- NA
```

## Data overview
This document describes the preprocessing and QC of microarray data from project MB189 tumor/normal tissues and organoids (Mirjana Kessler, Karen Hoffmann) using a single-color hybridization. 
Micro arrays used had design Agilent 014850 (4x44k Whole Human Genome v1). 

### Samples 
```{r,  results='asis', warning=FALSE}
cols_to_exclude = c("filename", "Sample.Number")
ed_filtered = ed[,!colnames(ed) %in% cols_to_exclude]
rownames(ed_filtered) <- NULL
#print.xtable(xtable(ed_filtered,display=rep("s",ncol(ed)-length(cols_to_exclude)+1), align=paste(c(rep("|l",ncol(ed)-length(cols_to_exclude)+1),"|"), collapse="")), type="html", file="" , include.rownames=F)
kable(ed_filtered)
```

## Raw intensity data

### Excluded probes (non-uniform feature or population outlier)

```{r, barplot_excl_probes, fig.width=10, fig.height=8}
par(mar=c(12,4,4,2))
ex_cnt = apply(intensities$probe_exclude,2,sum)
barplot(ex_cnt, las=2, names.arg = ed[names(ex_cnt), "ShortID"], main = "number of excluded probes/sample")
```

The array *Tissue Tumor 2* shows optical artefacts generating bias in a larger number of probes. 

### Intensity distribution across samples
```{r, boxplot_raw,  fig.width=10, fig.height=8}
par(mfrow=c(2,1), mar=c(11,4,4,1))
ii = log2(intensities$E)
colnames(ii) = ed[colnames(ii),]$ShortID
boxplot(ii, las=2, main = "Raw FG intensities", ylim=c(0,20))
ii = log2(intensities$Eb)
colnames(ii) = ed[colnames(ii),]$ShortID
boxplot(ii, las=2, main = "Raw BG intensities", ylim=c(0,20))
par(mfrow=c(1,1))
plotDensities(intensities, group = ed[colnames(intensities$E),"ShortID"], legend="topright", main="Intensity distribution by experiment run ID")

```

There are around 1500-2500 probes with background equal to zero in the last 4 arrays (M1 and M1_Wnt comparison) - not clear why this happened. Since we anyway will estimate the background from all probes using a statistical model below we will ignore this issue. 

### Correlation of raw intensities across samples
```{r, heatmap_cor_raw,  fig.width=12, fig.height=8}
raw_cor = cor(log2(intensities$E),method="spearman", use="pairwise")
colnames(raw_cor) = ed[colnames(raw_cor),]$ShortID
#rownames(raw_cor) = ed[rownames(raw_cor),]$sample_ID
pheatmap(raw_cor, cluster_rows = F, cluster_cols=F)
```

## Preprocessing

Data will be preprocessed as follows:

* Background for each array will be corrected using the "normexp" method (Ritchie et al 2007, Bioinformatics, p. 2700-07) and an offset of 15.
* Between-array normalization will be perfomed using the quantile method (Bolstad et al 2003, Bioinformatics, p. 185).

## Background correction
```{r, bg_correction, results='hide'}
bg_corrected = backgroundCorrect(intensities,method="normexp",offset=15)
```

### Intensity distribution across samples after background correction
```{r, boxplot_fg_bg_corrected,fig.width=10, fig.height=8}
par(mar=c(12,4,4,1))
boxplot(log2(bg_corrected$E), las=2,  names = ed[colnames(bg_corrected$E), "ShortID"], main = "BG-corrected FG intensities", ylim=c(-2,20))
```


### Normalization

```{r, normalization, results='hide'}
normalized = normalizeBetweenArrays(bg_corrected,method="quantile")
```

#### Intensity distribution across samples after BG correction and normalization
```{r, boxplot_fg_normalized,fig.width=10, fig.height=8}
boxplot(log2(normalized$E), las=2, names = ed[colnames(normalized$E), "ShortID"],main = "Normalized FG intensities", ylim=c(-2,20))
```

#### Correlation of normalized intensities across samples

```{r, heatmap_cor_normalized,  fig.width=12, fig.height=8}
norm_cor = cor(log2(normalized$E),method="spearman", use="pairwise")
colnames(norm_cor) = ed[colnames(norm_cor),]$sample_ID
rownames(norm_cor) = ed[rownames(norm_cor),]$ShortID
pheatmap(norm_cor)
```

## Primary Component Analysis

```{r, PCA,  fig.width=8, fig.height=8}
norm_exp = normalized$E
NA_rows = apply(norm_exp,1,function(x) sum(is.na(x)))
pca = prcomp(t(norm_exp[NA_rows==0,]))
imp = summary(pca)$importance
cp = palette(rainbow(8))
plot(pca$x[,1], pca$x[,2], type="p", xlab=paste("1st PC (",round(imp[2,1],2)*100 ,"% var explained)",sep="" ), ylab=paste("2nd PC (",round(imp[2,2],2)*100 ,"% var explained)",sep="" ), main="PCA on normalized expression data", xlim=c(min(pca$x[,1])*1.5,max(pca$x[,1])*4))
text(pca$x[,1],pca$x[,2],labels=ed[colnames(normalized$E),]$ShortID, col=cp[as.numeric(as.factor(ed[colnames(normalized$E),]$Condition))], adj=-0.05)
abline(h=0)
abline(v=0)
```

## Control probes

The following control probes exist on the arrays used in this experiment:

* Corner associated (used for orientation purposes during scanning)
  * Bright corner 
  * Dark corner 
* Negative controls
  * 3xSLv1 (hairpin probe that does not hybridize well with any possible RNA)
* Positive controls
  * Human GAPDH and PGK1 probes
  * Deletion stringency probes (DCP, probe with varying number of insertions/changes with respect to reference; the number after the "_" denotes the number of differences to the reference which should correlate with lower expression)
  * E1A_r60: spike-in probes with concentrations that should cover the whole dynamic range of the array

There are a few other expression probes that are used by Agilent's feature extraction/QC pipeline. 

```{r, qc_probes,  fig.width=12, fig.height=3}
control_probes = which(intensities$genes$ControlType!=0)
cp_data = intensities$E[control_probes,]
cp_names = intensities$genes[control_probes,]
selected_controls = ifelse(substr(cp_names$ProbeName,1,4)=="ERCC",F,T)

# control probes
for (i in 1:ncol(cp_data)) {
  boxplot(log2(cp_data[selected_controls,i]) ~ factor(cp_names$ProbeName[selected_controls]),las=2, main=paste("Sample",i), outline=F)
}
```

## Number of detected probes
Signal for probes will be considered detected if the expression of a probe is above the 95% quantile of expression in negative control probes. 
In a typical experiment one would expect about 50% of all genes to be detectable.

```{r, expr_count}
neg95 <- apply(normalized$E[normalized$genes$ControlType==-1,],2,function(x) quantile(x,p=0.95, na.rm=T))
cutoff <- matrix(1.1*neg95,nrow(normalized),ncol(normalized),byrow=TRUE)
isexpr <- rowSums(normalized$E > cutoff, na.rm=T) >= 2
# about 50% of genes should be detected
#table(isexpr)
```

```{r, number_expressed,fig.width=10, fig.height=8}
isexpr_sample <- colSums(normalized$E > cutoff, na.rm=T)
sample_order = order(ed[names(isexpr_sample),]$ShortID)
isexpr_sample_ordered = isexpr_sample[sample_order]
nn = ed[names(isexpr_sample),]$ShortID[sample_order]
par(mar=c(15,4,4,2))
barplot(isexpr_sample_ordered, main="number of detected probes per sample", las=2, names.arg=nn)
par(mar=c(8,4,4,2))
```

```{r, saving}
save(ed, intensities, normalized, file="../../Data/Processed/Tumor_organoids_micro_array_preprocessed_data.Rdata")
```

# Software versions

```{r}
sessionInfo()
```

